File: src/analysis/env_builder/pattern_helper.ml
(*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *)

module Ast = Flow_ast
open Ast.Pattern

type selector =
  | Elem of {
      index: int;
      has_default: bool;
    }
  | Prop of {
      prop: string;
      prop_loc: ALoc.t;
      has_default: bool;
    }
  | Computed of {
      expression: (ALoc.t, ALoc.t) Ast.Expression.t;
      has_default: bool;
    }
  | ObjRest of {
      used_props: string list;
      after_computed: bool;
    }
  | ArrRest of int
  | Default

type binding =
  | Root
  | Rest
  | Select of {
      selector: selector;
      parent: ALoc.t * binding;
    }

let array_element (parent_loc, bind) index direct_default =
  let selector = Elem { index; has_default = direct_default <> None } in
  Select { selector; parent = (parent_loc, bind) }

let array_rest_element (parent_loc, bind) i =
  let selector = ArrRest i in
  Select { selector; parent = (parent_loc, bind) }

let object_named_property (parent_loc, bind) prop_loc x direct_default =
  let has_default = direct_default <> None in
  let selector = Prop { prop = x; prop_loc; has_default } in
  Select { selector; parent = (parent_loc, bind) }

let object_computed_property (parent_loc, bind) e direct_default =
  let selector = Computed { expression = e; has_default = direct_default <> None } in
  Select { selector; parent = (parent_loc, bind) }

let object_rest_property (parent_loc, bind) xs has_computed =
  let selector = ObjRest { used_props = xs; after_computed = has_computed } in
  Select { selector; parent = (parent_loc, bind) }

let object_property (parent_loc, bind) xs key direct_default =
  let open Ast.Pattern.Object in
  match key with
  | Property.Identifier (loc, { Ast.Identifier.name = x; comments = _ }) ->
    let bind = object_named_property (parent_loc, bind) loc x direct_default in
    (bind, x :: xs, false)
  | Property.Literal (loc, { Ast.Literal.value = Ast.Literal.String x; _ }) ->
    let bind = object_named_property (parent_loc, bind) loc x direct_default in
    (bind, x :: xs, false)
  | Property.Computed (_, { Ast.ComputedKey.expression; comments = _ }) ->
    let bind = object_computed_property (parent_loc, bind) expression direct_default in
    (bind, xs, true)
  | Property.Literal (_, _) -> (bind, xs, false)

let identifier acc bind (name_loc, { Ast.Identifier.name; _ }) = SMap.add name (name_loc, bind) acc

let rec fold_pattern acc bind (ploc, p) =
  match p with
  | Array { Array.elements; annot = _; comments = _ } -> array_elements acc (ploc, bind) elements
  | Object { Object.properties; annot = _; comments = _ } ->
    object_properties acc (ploc, bind) properties
  | Identifier { Identifier.name = id; optional = _; annot = _ } -> identifier acc bind id
  | Expression _ -> acc

and array_elements acc (parent_loc, bind) elts =
  let open Ast.Pattern.Array in
  Base.List.fold
    ~init:(0, acc)
    ~f:(fun (i, acc) elt ->
      let acc =
        match elt with
        | Hole _ -> acc
        | Element (_, { Element.argument = p; default = d }) ->
          let bind = array_element (parent_loc, bind) i d in
          fold_pattern acc bind p
        | RestElement (_, { Ast.Pattern.RestElement.argument = p; comments = _ }) ->
          let bind = array_rest_element (parent_loc, bind) i in
          fold_pattern acc bind p
      in
      (i + 1, acc))
    elts
  |> snd

and object_properties =
  let open Ast.Pattern.Object in
  let prop acc (parent_loc, bind) xs has_computed p =
    match p with
    | Property (_, { Property.key; pattern = p; default = d; shorthand = _ }) ->
      let (bind, xs, has_computed') = object_property (parent_loc, bind) xs key d in
      (fold_pattern acc bind p, xs, has_computed || has_computed')
    | RestElement (_, { Ast.Pattern.RestElement.argument = p; comments = _ }) ->
      let bind = object_rest_property (parent_loc, bind) xs has_computed in
      (fold_pattern acc bind p, xs, false)
  in
  let rec loop acc bind xs has_computed = function
    | [] -> acc
    | p :: ps ->
      let (acc, xs, has_computed) = prop acc bind xs has_computed p in
      loop acc bind xs has_computed ps
  in
  (fun acc bind ps -> loop acc bind [] false ps)

let bindings_of_params (_, { Ast.Function.Params.params; rest; _ }) =
  let acc =
    Base.List.fold params ~init:SMap.empty ~f:(fun acc (_, { Ast.Function.Param.argument; _ }) ->
        fold_pattern acc Root argument
    )
  in
  Base.Option.fold rest ~init:acc ~f:(fun acc (_, { Ast.Function.RestParam.argument; _ }) ->
      fold_pattern acc Rest argument
  )

let bindings_of_pattern pattern = fold_pattern SMap.empty Root pattern

File: rust_port/crates/flow_env_builder/src/pattern_helper.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::BTreeMap;
use std::rc::Rc;

use dupe::Dupe;
use flow_common::js_number::ecma_string_of_float;
use flow_common::js_number::is_float_safe_integer;
use flow_data_structure_wrapper::smol_str::FlowSmolStr;
use flow_parser::ast::expression::Expression;
use flow_parser::ast::function;
use flow_parser::ast::pattern;
use flow_parser::ast::pattern::Pattern;

use crate::selector::Selector;

#[derive(Debug, Clone)]
pub enum Binding<L: Dupe, T: Dupe> {
    Root,
    Rest,
    Select {
        selector: Selector<L, T>,
        parent: (L, Rc<Binding<L, T>>),
    },
}

pub fn array_element<L: Dupe, T: Dupe>(
    parent_loc: L,
    bind: Rc<Binding<L, T>>,
    index: usize,
    direct_default: bool,
) -> Rc<Binding<L, T>> {
    let selector = Selector::Elem {
        index,
        has_default: direct_default,
    };
    Rc::new(Binding::Select {
        selector,
        parent: (parent_loc, bind),
    })
}

pub fn array_rest_element<L: Dupe, T: Dupe>(
    parent_loc: L,
    bind: Rc<Binding<L, T>>,
    index: usize,
) -> Rc<Binding<L, T>> {
    let selector = Selector::ArrRest(index);
    Rc::new(Binding::Select {
        selector,
        parent: (parent_loc, bind),
    })
}

pub fn object_named_property<L: Dupe, T: Dupe>(
    parent_loc: L,
    bind: Rc<Binding<L, T>>,
    prop_loc: L,
    name: FlowSmolStr,
    has_default: bool,
) -> Rc<Binding<L, T>> {
    let selector = Selector::Prop {
        prop: name,
        prop_loc,
        has_default,
    };
    Rc::new(Binding::Select {
        selector,
        parent: (parent_loc, bind),
    })
}

pub fn object_computed_property<L: Dupe, T: Dupe>(
    parent_loc: L,
    bind: Rc<Binding<L, T>>,
    expression: Expression<L, T>,
    has_default: bool,
) -> Rc<Binding<L, T>> {
    let selector = Selector::Computed {
        expression,
        has_default,
    };
    Rc::new(Binding::Select {
        selector,
        parent: (parent_loc, bind),
    })
}

pub fn object_rest_property<L: Dupe, T: Dupe>(
    parent_loc: L,
    bind: Rc<Binding<L, T>>,
    used_props: &[FlowSmolStr],
    after_computed: bool,
) -> Rc<Binding<L, T>> {
    let selector = Selector::ObjRest {
        used_props: used_props.iter().map(|s| s.dupe()).collect(),
        after_computed,
    };
    Rc::new(Binding::Select {
        selector,
        parent: (parent_loc, bind),
    })
}

pub fn object_property<L: Dupe>(
    parent_loc: L,
    bind: Rc<Binding<L, L>>,
    mut used_props: Vec<FlowSmolStr>,
    key: &pattern::object::Key<L, L>,
    has_default: bool,
) -> (Rc<Binding<L, L>>, Vec<FlowSmolStr>, bool) {
    match key {
        pattern::object::Key::Identifier(id) => {
            let name = id.name.dupe();
            let bind =
                object_named_property(parent_loc, bind, id.loc.dupe(), name.dupe(), has_default);
            used_props.push(name);
            (bind, used_props, false)
        }
        pattern::object::Key::StringLiteral((loc, lit)) => {
            let name = lit.value.dupe();
            let bind =
                object_named_property(parent_loc, bind, loc.dupe(), name.dupe(), has_default);
            used_props.push(name);
            (bind, used_props, false)
        }
        pattern::object::Key::Computed(computed) => {
            let bind = object_computed_property(
                parent_loc,
                bind,
                computed.expression.clone(),
                has_default,
            );
            (bind, used_props, true)
        }
        pattern::object::Key::NumberLiteral((loc, lit)) => {
            if is_float_safe_integer(lit.value) {
                let name = FlowSmolStr::from(ecma_string_of_float(lit.value));
                let bind =
                    object_named_property(parent_loc, bind, loc.dupe(), name.dupe(), has_default);
                used_props.push(name);
                (bind, used_props, false)
            } else {
                (Rc::new(Binding::Root), used_props, false)
            }
        }
        pattern::object::Key::BigIntLiteral(_) => (Rc::new(Binding::Root), used_props, false),
    }
}

fn identifier<L: Dupe + Ord>(
    mut acc: BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)>,
    bind: Rc<Binding<L, L>>,
    name_loc: L,
    name: FlowSmolStr,
) -> BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)> {
    acc.insert(name, (name_loc, bind));
    acc
}

pub fn fold_pattern<L: Ord + Dupe>(
    acc: BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)>,
    bind: Rc<Binding<L, L>>,
    pattern: &Pattern<L, L>,
) -> BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)> {
    match pattern {
        Pattern::Array { loc, inner, .. } => array_elements(acc, loc.dupe(), bind, &inner.elements),
        Pattern::Object { loc, inner, .. } => {
            object_properties(acc, loc.dupe(), bind, &inner.properties)
        }
        Pattern::Identifier { inner, .. } => {
            let name_loc = inner.name.loc.dupe();
            let name = inner.name.name.dupe();
            identifier(acc, bind, name_loc, name)
        }
        Pattern::Expression { .. } => acc,
    }
}

fn array_elements<L: Ord + Dupe>(
    mut acc: BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)>,
    parent_loc: L,
    bind: Rc<Binding<L, L>>,
    elements: &[pattern::array::Element<L, L>],
) -> BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)> {
    for (i, elt) in elements.iter().enumerate() {
        acc = match elt {
            pattern::array::Element::Hole(_) => acc,
            pattern::array::Element::NormalElement(elem) => {
                let has_default = elem.default.is_some();
                let child_bind = array_element(parent_loc.dupe(), bind.dupe(), i, has_default);
                fold_pattern(acc, child_bind, &elem.argument)
            }
            pattern::array::Element::RestElement(rest) => {
                let child_bind = array_rest_element(parent_loc.dupe(), bind.dupe(), i);
                fold_pattern(acc, child_bind, &rest.argument)
            }
        };
    }
    acc
}

fn object_properties<L: Ord + Dupe>(
    mut acc: BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)>,
    parent_loc: L,
    bind: Rc<Binding<L, L>>,
    properties: &[pattern::object::Property<L, L>],
) -> BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)> {
    let mut used_props: Vec<FlowSmolStr> = Vec::new();
    let mut has_computed = false;

    for prop in properties {
        match prop {
            pattern::object::Property::NormalProperty(normal) => {
                let prop_has_default = normal.default.is_some();
                let (prop_bind, new_used_props, new_has_computed) = object_property(
                    parent_loc.dupe(),
                    bind.dupe(),
                    used_props,
                    &normal.key,
                    prop_has_default,
                );
                used_props = new_used_props;
                has_computed = has_computed || new_has_computed;
                acc = fold_pattern(acc, prop_bind, &normal.pattern);
            }
            pattern::object::Property::RestElement(rest) => {
                let rest_bind =
                    object_rest_property(parent_loc.dupe(), bind.dupe(), &used_props, has_computed);
                acc = fold_pattern(acc, rest_bind, &rest.argument);
                has_computed = false;
            }
        }
    }
    acc
}

pub fn bindings_of_params<L: Ord + Dupe>(
    params: &function::Params<L, L>,
) -> BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)> {
    let mut acc = BTreeMap::new();
    for param in params.params.iter() {
        match param {
            function::Param::RegularParam { argument, .. } => {
                acc = fold_pattern(acc, Rc::new(Binding::Root), argument);
            }
            function::Param::ParamProperty { .. } => {
                // Skip param properties - they are not supported
            }
        }
    }

    if let Some(rest) = &params.rest {
        acc = fold_pattern(acc, Rc::new(Binding::Rest), &rest.argument);
    }

    acc
}

pub fn bindings_of_pattern<L: Ord + Dupe>(
    pattern: &Pattern<L, L>,
) -> BTreeMap<FlowSmolStr, (L, Rc<Binding<L, L>>)> {
    fold_pattern(BTreeMap::new(), Rc::new(Binding::Root), pattern)
}
